core/test/base/segmented_range.cpp

```diff
@@ -0,0 +1,136 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <iterator>
+#include <numeric>
+#include <vector>
+
+
+#include <gtest/gtest.h>
+
+
+#include "core/base/segmented_range.hpp"
+
+
+TEST(SegmentedRange, Works)
+{
+    std::vector<int> begins{3, 1, 4, 9};
+    std::vector<int> ends{3, 10, 6, 10};
+    std::vector<std::vector<int>> result_indices(begins.size());
+    gko::segmented_range<int> range{begins.data(), ends.data(),
+                                    static_cast<int>(begins.size())};
+
+    for (auto row : gko::irange<int>(begins.size())) {
```

Comment:
nit:
```suggestion
    for (auto row : gko::irange<int>(range.num_segments())) {
```
since I think this shows better the usage of the range.
Same for the rest of the file.